lib/Dialect/Polynomial/Transforms/NTTRewrites.td

```diff
@@ -0,0 +1,68 @@
+#ifndef LIB_DIALECT_POLYNOMIAL_TRANSFORMS_NTTREWRITES_TD_
+#define LIB_DIALECT_POLYNOMIAL_TRANSFORMS_NTTREWRITES_TD_
+
+include "lib/DRR/Utils.td"
+include "lib/Dialect/Polynomial/IR/PolynomialOps.td"
+include "mlir/Dialect/Arith/IR/ArithOps.td"
+include "mlir/IR/PatternBase.td"
+
+def GetRingAttr : NativeCodeCall<
+      "(dyn_cast<PolynomialType>($0.getType())).getRing()">;
+
+def InputTensorType : NativeCodeCall<
+      "RankedTensorType::get({$0.ideal().getDegree()},"
+      " IntegerType::get($_builder.getContext(),"
+      "  ($0.coefficientModulus() - 1).getActiveBits()),"
+      " $0)">;
+
+def IntermediateTensorType : NativeCodeCall<
+      "RankedTensorType::get({$0.ideal().getDegree()},"
+      " IntegerType::get($_builder.getContext(),"
+      "  2 * ($0.coefficientModulus() - 1).getActiveBits()),"
+      " $0)">;
+
+def CreateCModConstant : NativeCodeCall<
+      "$_builder.create<arith::ConstantOp>($0.getLoc(), $2,"
+      " DenseElementsAttr::get($2,"
+      "  $1.coefficientModulus().sextOrTrunc($2.getElementTypeBitWidth())))">;
+
+// Computes (n & (n - 1)) == 0. Can only pass in symbols to the DRR constraint
+// so unfortunately need to extract the ring and degree each time...
+def HasDegreePowerOfTwo : Constraint<
```

Comment:
Might be slightly simpler to wrap it in an APInt and call `APInt::isPowerOf2`, at least since you know the degree has <= 64 bits of precision.